Rewrite Generative Results (#36)
greenelab/text_mined_hetnet_manuscript: content/04.results.md

## Results

### Generative Model Using Randomly Sampled Label Functions

Creating label functions is a labor intensive process that can take days to accomplish.
We sought to accelerate this process by measuring the extent to which label functions can be reused.

> Our hypothesis is that certain edge types share similar linguistic features such as keywords and/or sentence structure.

Reviewer: ```suggestion
Our hypothesis was that certain edge types share similar linguistic features such as keywords and/or sentence structure.
```

This shared characteristic would make certain edge types amenable to label function reuse.

> We incorporated a sampling approach to test this hypothesis on an individual level (edge vs edge) as well as a global level (collective pool of sources). 

Reviewer: ```suggestion
We designed a set of experiments to test this hypothesis on an individual level (edge vs edge) as well as a global level (collective pool of sources). 
```

We report results in terms of AUROC (Figures {@fig:auroc_gen_model_test_set} and {@fig:auroc_grabbag_gen_model_test_set}) and AUPR (Supplemental Figure {@fig:aupr_gen_model_test_set} and {@fig:aupr_grabbag_gen_model_test_set}).

> Performance increases when edge-specific label functions are added to an edge-specific baseline model, while label function reusability shows modest results.

Reviewer: Refer to the figures in the first sentence about their results. So the first figure reference would go into this sentence and the one before would be removed.

> The quintessential example of the first trend is the Compound treats Disease (CtD) edge type, where edge-specific label functions completely dominates the edge-mismatch label functions.

Reviewer: ```suggestion
The quintessential example of the overarching trend is the Compound treats Disease (CtD) edge type, where edge-specific label functions always outperformed transferred label functions.
```

> A mildly similar trend can be seen within the Compound binds Gene (CbG) and Gene interacts Gene (GiG) edge types as edge-specific label functions dominate performance.

Reviewer: ```suggestion
```

> CbG and GiG show positive signs of label function transferability. 

Reviewer: ```suggestion
However, there are hints of label function transferability for selected edge types and label function sources. 
```

Performance increases as more CbG label functions are incorporated to the GiG baseline model and vise-versa.
This suggests that sentences for GiG and CbG may share similar linguistic features or terminology that allows for label functions to be reused.

> Perplexingly, edge-specific Disease associates Gene (DaG) label functions fail to outcompete other label function sources.

Reviewer: ```suggestion
Edge-specific Disease associates Gene (DaG) label functions did not improve performance over label functions drawn from other edge types.
```

> Overall, only CbG and GiG show significant signs of reusability which suggests label functions could be shared between the two edge types.

Reviewer: Well - there is one other interesting case if I'm understanding this correctly - DaG can use any label functions. It's the universal acceptor for label functions. Did I get that right?

Reviewer: So really there are two cases - one that is reciprocal high performance and another where you can use basically anything.

![

> Edge-specific label functions improves performance over edge-mismatch label functions.

Reviewer: ```suggestion
Edge-specific label functions are better performing than edge-mismatch label functions but certain mismatch situations show signs of successful transfer.
```

Each line plot header depicts the edge type the generative model is trying to predict, while the colors represent the source of label functions.
For example orange represents sampling label functions designed to predict the Compound treats Disease (CtD) edge type.
The x axis shows the number of randomly sampled label functions being incorporated onto the database only baseline model (point at 0).
The y axis shows area under the receiver operating curve (AUROC).
Each point on the plot shows the average of 50 sample runs, while the error bars show the 95% confidence intervals of all runs.
The baseline and “All” data points consist of sampling from the entire fixed set of label functions.

> This makes the error bars appear flat.

Reviewer: ```suggestion
```

](https://raw.githubusercontent.com/danich1/snorkeling/86037d185a299a1f6dd4dd68605073849c72af6f/figures/label_sampling_experiment/transfer_test_set_auroc.png){#fig:auroc_gen_model_test_set}

> Utilizing all label functions at once generally hinders performance of the generative model.

Reviewer: ```suggestion
We found that sampling from all label function sources at once usually underperformed relative to edge-specific label functions (Figure {#fig:auroc_grabbag_gen_model_test_set}).
```

> Typically as more label functions are being sampled the gap between edge-specific sources and the all category widen.

Reviewer: ```suggestion
As more label functions were sampled, the gap between edge-specific sources and all sources widened.
```

CbG is a prime example of this trend (Figure {#fig:auroc_grabbag_gen_model_test_set}), while CtD and GiG show a similar but milder trend.

> We expected this trend to occur, since most edge-mismatch label functions showed a deteriorating trend in Figure {#fig:auroc_gen_model_test_set}.

Reviewer: ```suggestion
```

> An exception to the shared consensus is DaG where the pooled set of label functions improved performance over the edge-specific case.

Reviewer: ```suggestion
DaG was the exception to the general rule: the pooled set of label functions improved performance over the edge-specific ones, which aligns with the previously observed results for individual edge types (Figure {#fig:auroc_gen_model_test_set}).
```

The decreasing trend when pooling all label functions supports the notion that label functions cannot easily transfer between edge types (exception being CbG on GiG and vise versa).

![
Using all label functions generally hinders generative model performance.
Each line plot header depicts the edge type the generative model is trying to predict, while the colors represent the source of label functions.
For example, orange represents sampling label functions designed to predict the Compound treats Disease (CtD) edge type.
The x axis shows the number of randomly sampled label functions being incorporated onto the database only baseline model (point at 0).
The y axis shows area under the receiver operating curve (AUROC).
Each point on the plot shows the average of 50 sample runs, while the error bars show the 95% confidence intervals of all runs.
The baseline and “All” data points consist of sampling from the entire fixed set of label functions.

> This makes the error bars appear flat.

Reviewer: ```suggestion
```

### Generative Model Using Randomly Sampled Label Functions
![

> Plot of AUROC scores for each generative model trained on randomly sampled label functions.

Reviewer: As with other figures - give me a short summary of your takeaway in the first line of the legend.

> The headers depict the relationship each model is trying to predict and the colored points are the edge type specific sources from which each label function is sampled.

Reviewer: ```suggestion
The headers depict the target relationship and the colored of each point denotes the sources from which functions are sampled.
```

> The 0th point represents our baseline model which only uses only distant supervision label functions.

Reviewer: ```suggestion
The point at 0 is a baseline model that only uses distant supervision label functions.
```

> ](https://raw.githubusercontent.com/danich1/snorkeling/86037d185a299a1f6dd4dd68605073849c72af6f/figures/label_sampling_experiment/transfer_test_set_auroc.png){#fig:auroc_gen_model_test_set}

Reviewer: What do error bars depict? Should be in legend.

![

> Plot of AUPR scores for each generative model trained on randomly sampled label functions.

Reviewer: Same comments as above, but make those changes here too.

The headers depict the relationship each model is trying to predict and the colored points represent the edge type specific sources from which each label function is sampled.
The 0th point represents our baseline model which only uses distant supervision label functions.
](https://raw.githubusercontent.com/danich1/snorkeling/86037d185a299a1f6dd4dd68605073849c72af6f/figures/label_sampling_experiment/transfer_test_set_aupr.png){#fig:aupr_gen_model_test_set}

> We added randomly sampled label functions to a baseline for each edge type to evaluate the feasibility of label function re-use.

Reviewer: First sentence: what is this experiment testing. Need the context here. The figures should come after the first paragraph where they are discussed, not before.

Our baseline model consisted of a generative model trained with only edge-specific distant supervision label functions.

> We reported the results in AUROC and AUPR (Figures {@fig:auroc_gen_model_test_set} and {@fig:aupr_gen_model_test_set}; Supplemental Figures {@fig:auroc_gen_model_tune_set} and {@fig:auroc_gen_model_tune_set}).  

Reviewer: Make an assertion about performance that is supported by the figures for the first time you reference them. Similar to including the takeaway point in the first sentence of the legend. It makes your paper easier to read :)

> Performance increases as the number of edge-specific label functions is added to the baseline model (Figures {@fig:auroc_gen_model_test_set} and {@fig:aupr_gen_model_test_set}). 

Reviewer: ```suggestion
Performance increases when edge-specific label functions are added to an edge-specific baseline model trained with distant supervision (Figures {@fig:auroc_gen_model_test_set} and {@fig:aupr_gen_model_test_set}). 
```
You could probably move either AUPR or AUROC to the supplement. I'm not sure what having both figures here is adding, unless there's a difference.

The Compound treats Disease (CtD) edge type is a quintessential example of this trend.

> The baseline model starts off with an AUROC score of 52% and an AUPRC of 29%, then performance increases to 76% and 48% respectively as more CtD related label functions are included. 

Reviewer: Don't include specific numbers in writing when the values are in the figure unless it adds something above and beyond the figure.

> The baseline model starts off with an AUROC score of 52% and an AUPRC of 29%, then performance increases to 76% and 48% respectively as more CtD related label functions are included. 

Reviewer: Don't include specific numbers in writing when the values are in the figure unless it adds something above and beyond the figure.

Gene interacts Gene (GiG), Disease associates Gene (DaG) and Compound binds Gene (CbG) edge types share a similar trend.
Baseline performance starts off with an AUROC of 57% for DaG, 78% for CbG and 66% for GiG, then increases to 62% for DaG, 81% for CbG and 75% for GiG.
Similarly with AUPR, the baseline starts off at 43% for DaG, 29% for CbG and 30% for GiG, then increases to 44%, 30% and 35% respectively.

We based our initial hypothesis on the idea that certain edge types would share related linguistic features and terminology which would be amenable to label function transfer.

> For example, sentences for edges CbG and GiG typically describe physical interactions whether it be between a compound and a gene or a pair of two protein produced from genes.

Reviewer: Train wifi is dying on me but I couldn't recall if this feature was retained or went away. Since I closed the figure and now can't re-load it, is this still the case?

Reviewer: Ok - it's working again - is it weird that DAG gets better from every edge type?

We observed performance increases in terms of both AUROC and AUPR when using label functions from the GiG edge type to predict CbG edges (Figures {@fig:auroc_gen_model_test_set} and {@fig:aupr_gen_model_test_set}).
Likewise when using CbG related label function to predict the GiG edge type; however, in both cases GiG related label functions out performed the CbG related label functions.

> Regarding DaG, each label function source had similar performance when predicting the test set (Figures {@fig:auroc_gen_model_test_set} and {@fig:aupr_gen_model_test_set}); however the difference is more apparent when predicting the tune set (Supplemental Figures {@fig:auroc_gen_model_tune_set} and {@fig:auroc_gen_model_tune_set}).

Reviewer: If you're going to bring in the tune set it'd be helpful for me to understand why. Is there an interpretation that the reader needs to take away here? Why not just focus on the test set, where we expect the estimate of performance to be accurate?

We pooled together all non-database label functions for each edge type.
We randomly sampled label functions from the combined pool, then trained a generative model to predict each edge type.
Performance is report in terms of AUROC and AUPR (Figures {@fig:auroc_grabbag_gen_model_test_set} and {@fig:aupr_grabbag_gen_model_test_set}; Supplemental Figures @fig:auroc_grabbag_gen_model_tune_set} and @fig:aupr_grabbag_gen_model_tune_set}).
General consensus for CbG, GiG and CtD edges is that using all label functions hurt performance.
The gap between edge specific sources and the all category widens as the number of sampled label functions increases.
The only exception to this is DaG where performance increases (57% to 65% AUROC and 43% to 47% AUPR) as more and more label functions are added to the baseline model.

Overall, only CbG and GiG show significant signs of transferability between each other for both Tune and Test set.

> DaG shows conflicting results between the tune and test set, when it comes to sampling individual off-edge label functions; however, when pooling every label function together performance increases. 

Reviewer: I think that when you see a difference between tune and test you have to trust test, right? We don't expect tune to be an accurate estimate of performance.
